extensions/notebook/src/book/bookModel.ts

```diff
@@ -0,0 +1,211 @@
+			let tableOfContentPaths = await glob(p, { deep: maxDepth });
+			this._tableOfContentPaths = this._tableOfContentPaths.concat(tableOfContentPaths);
+			let bookOpened: boolean = this._tableOfContentPaths.length > 0;
+			vscode.commands.executeCommand('setContext', 'bookOpened', bookOpened);
+		} catch (error) {
+			console.log(error);
+		}
+	}
+
+	public async readBooks(): Promise<BookTreeItem[]> {
+		for (const contentPath of this._tableOfContentPaths) {
+			let root = path.dirname(path.dirname(contentPath));
+			try {
+				let fileContents = await fsPromises.readFile(path.join(root, '_config.yml'), 'utf-8');
+				const config = yaml.safeLoad(fileContents.toString()); //fs.readFileSync(path.join(root, '_config.yml'), 'utf-8')
+				fileContents = await fsPromises.readFile(contentPath, 'utf-8');
+				const tableOfContents = yaml.safeLoad(fileContents.toString());
+				let book = new BookTreeItem({
+					title: config.title,
+					root: root,
+					tableOfContents: { sections: this.parseJupyterSection(tableOfContents) },
+					page: tableOfContents,
+					type: BookTreeItemType.Book,
+					treeItemCollapsibleState: vscode.TreeItemCollapsibleState.Expanded,
+					isUntitled: this.openAsUntitled,
+				},
+					{
+						light: this._extensionContext.asAbsolutePath('resources/light/book.svg'),
+						dark: this._extensionContext.asAbsolutePath('resources/dark/book_inverse.svg')
+					}
+				);
+				this._bookItems.push(book);
+			} catch (e) {
+				let error = e instanceof Error ? e.message : e;
+				vscode.window.showErrorMessage(error);
+			}
+		}
+		return this._bookItems;
+	}
+
+	public get bookItems(): BookTreeItem[] {
+		return this._bookItems;
+	}
+
+	public getSections(tableOfContents: IJupyterBookToc, sections: IJupyterBookSection[], root: string): BookTreeItem[] {
+		let notebooks: BookTreeItem[] = [];
+		for (let i = 0; i < sections.length; i++) {
+			if (sections[i].url) {
+				if (sections[i].external) {
+					let externalLink = new BookTreeItem({
+						title: sections[i].title,
+						root: root,
+						tableOfContents: tableOfContents,
+						page: sections[i],
+						type: BookTreeItemType.ExternalLink,
+						treeItemCollapsibleState: vscode.TreeItemCollapsibleState.Collapsed,
+						isUntitled: this.openAsUntitled
+					},
+						{
+							light: this._extensionContext.asAbsolutePath('resources/light/link.svg'),
+							dark: this._extensionContext.asAbsolutePath('resources/dark/link_inverse.svg')
+						}
+					);
+
+					notebooks.push(externalLink);
+				} else {
+					let pathToNotebook = path.join(root, 'content', sections[i].url.concat('.ipynb'));
+					let pathToMarkdown = path.join(root, 'content', sections[i].url.concat('.md'));
+					// Note: Currently, if there is an ipynb and a md file with the same name, Jupyter Books only shows the notebook.
+					// Following Jupyter Books behavior for now
+					if (fs.existsSync(pathToNotebook)) {
+						let notebook = new BookTreeItem({
```

Comment:
Don't use sync functions - can you make this function async like Amir suggested?

Reply:
This will also break the builds due to our new no sync methods rule.

Reply:
That's only applied to the core - not to extensions (but is still something the extensions should be moving to doing)